trie/triedb/pathdb/disklayer.go

```diff
@@ -0,0 +1,259 @@
+	return newDiffLayer(dl, stateRoot, id, blockNumber, nodes, states)
+}
+
+// commit merges the given bottom-most diff layer into the node buffer
+// and returns a newly constructed disk layer. Note the current disk
+// layer must be tagged as stale first to prevent re-access.
+func (dl *diskLayer) commit(bottom *diffLayer, force bool) (*diskLayer, error) {
+	dl.lock.Lock()
+	defer dl.lock.Unlock()
+
+	// Construct and store the state history first. If crash happens
+	// after storing the state history but without flushing the
+	// corresponding states(journal), the stored state history will
+	// be truncated in the next restart.
+	if dl.db.freezer != nil {
+		err := writeStateHistory(dl.db.freezer, bottom, dl.db.config.StateLimit)
+		if err != nil {
+			return nil, err
+		}
+	}
+	// Mark the diskLayer as stale before applying any mutations on top.
+	dl.stale = true
+
+	// Store the root->id lookup afterwards. All stored lookups are
+	// identified by the **unique** state root. It's impossible that
+	// in the same chain blocks are not adjacent but have the same
+	// root.
+	if dl.id == 0 {
+		rawdb.WriteStateID(dl.db.diskdb, dl.rootHash, 0)
+	}
+	rawdb.WriteStateID(dl.db.diskdb, bottom.root(), bottom.stateID())
+
+	// Construct a new disk layer by merging the nodes from the provided
+	// diff layer, and flush the content in disk layer if there are too
+	// many nodes cached.
+	ndl := newDiskLayer(bottom.rootHash, bottom.id, dl.db, dl.buffer.commit(bottom.nodes))
+	err := ndl.buffer.flush(ndl.db.diskdb, ndl.db.cleans, ndl.id, force)
+	if err != nil {
+		return nil, err
+	}
+	return ndl, nil
+}
+
+// revert applies the given state history and return a reverted disk layer.
+func (dl *diskLayer) revert(h *history, loader triestate.TrieLoader) (*diskLayer, error) {
+	if h.meta.root != dl.root() {
+		return nil, errUnexpectedHistory
+	}
+	// Reject if the provided state history is incomplete. It's due to
+	// a large construct SELF-DESTRUCT which can't be handled because
+	// of memory limitation.
+	if len(h.meta.incomplete) > 0 {
+		return nil, errors.New("incomplete state history")
+	}
+	if dl.id == 0 {
+		return nil, fmt.Errorf("%w: zero state id", errStateUnrecoverable)
+	}
+	// Apply the reverse state changes upon the current state. This must
+	// be done before holding the lock in order to access state in "this"
+	// layer.
+	nodes, err := triestate.Apply(h.meta.parent, h.meta.root, h.accounts, h.storages, loader)
+	if err != nil {
+		return nil, err
+	}
+	// Mark the diskLayer as stale before applying any mutations on top.
+	dl.lock.Lock()
+	defer dl.lock.Unlock()
+
+	dl.stale = true
+
+	// Revert embedded states in the node buffer first if it's not empty.
+	if !dl.buffer.empty() {
+		err := dl.buffer.revert(dl.db.diskdb, nodes)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// Reset the clean cache before mutating persistent state.
+		if dl.db.cleans != nil {
+			dl.db.cleans.Reset()
```

Comment:
dl.db.cleans is the cache relevant with *disk*, but not the entire diskLayer.

I agree the name of diskLayer is confusing, but actually the proper name
is the bottom-most layer or so.

If we revert data in buffer, we just revoke the changes happened in the
memory, but not in the disk. If we revert data in disk, we are actually 
revoke the changes happened in disk.

Since cache is corresponding to disk, whenever the persistent state 
in disk is reverted, we should reset the cache as well.